doc/sharedtable.md

```diff
@@ -0,0 +1,61 @@
+# sharedtable #
+
+```lua
+local t = ipc.sharedtable([tbl])
+```
+
+A `sharedtable` is a table that can be shared between threads.
+The constructor can take an optional `tbl` argument, which is a Lua table.
+When provided, `tbl` is used to initialize the `sharedtable`.
+
+A `sharedtable` can be shared between threads, either by passing it through
+a `ipc.workqueue` or as an argument to `ipc.map`.
+Here is an example using `ipc.map`:
+
+```lua
+local t = ipc.sharedtable({0})
+ipc.map(2, function(tab, threadid)
+   for i=1,100 do
+      if (i % 2) + 1 == threadid then
+         tab[1] = i
+      end
+   end
+end, t):join()
+
+for i=1,100 do
+   assert(t[i] == i)
+end
+```
+
+Indeed, the `sharedtable` instance `t` is shared between 2 worker threads and the main thread.
+
+Internally, the `sharedtable` is implemented by storing the table in a separate `lua_State`.
+Recall that each thread also has its own `lua_State`.
+And the way data is passed between threads is by serializing/deserializing the data from one `lua_State` to another.
+The `sharedtable` uses the same principle.
+By this I mean that all read and writes to the table (get/set) are implemented by (de)serializing between the calling thread's and the shared table's `lua_State`.
+
+The `sharedtable` is subject to some caveats.
+Some use-cases will still require a lock.
+For example:
+
+```lua
+
+local t = ipc.sharedtable()
+local m = ipc.mutex()
+ipc.map(10, function(tab, mutex)
+   for i=1,100 do
+      mutex:lock()
+      tab[i] = (tab[i] or 0) + 1
+      mutex:unlock()
+   end
+end, t, m):join()
+assert(#t == 100)
+for i=1,100 do
+   assert(t[i] == 10)
+end
+```
+
+In the above example, we want to atomically add 1 to the 100 first element of the table.
+Doing this requires a `mutex`.
+
```

Comment:
A very big caveat of `sharedtable` that should be highlighted is that, when you de-serialize, you're actually creating an unnamed variable in the thread's state. This causes some confusion, as in:
```lua
local t1 = {'key': {'subkey': 1}}
local t2 = {'key': 1}
local st1 = ipc.sharedtable(t1)
local st2 = ipc.sharedtable(t2)
st1.key.subkey = 2
st2.key = 2
assert(st2.key == 1) -- works because you set 2 in a local variable created by the call st1.key, which isn't transported back
assert(st2.key == 2) -- works because you set 2 in the sharedtable directly
```

This, of course, is solved if every table inside `sharedtable` is actually also a `sharedtable`, but it's important to highlight this as overwriting this behavior would require changes on lua itself.